File: 18_multiplication_linalg/poly.py
def poly_add(p, q):
    n = max(len(p), len(q))
    return [(p[i] if i < len(p) else 0) + (q[i] if i < len(q) else 0) for i in range(n)]

def poly_sub(p, q):
    n = max(len(p), len(q))
    return [(p[i] if i < len(p) else 0) - (q[i] if i < len(q) else 0) for i in range(n)]

def pad(poly, length):
    return poly + [0] * (length - len(poly))

def poly_multiply(p, q):
    n = max(len(p), len(q))
    if n == 1:
        return [p[0] * q[0]]

    m = (n + 1) // 2
    p = pad(p, 2 * m)
    q = pad(q, 2 * m)

    low_p, high_p = p[:m], p[m:]
    low_q, high_q = q[:m], q[m:]

    z0 = poly_multiply(low_p, low_q)
    z2 = poly_multiply(high_p, high_q)
    z1 = poly_multiply(poly_add(low_p, high_p), poly_add(low_q, high_q))
    z1 = poly_sub(poly_sub(z1, z0), z2)

    # combine results
    result = [0] * (4 * m)
    for i in range(len(z0)): result[i] += z0[i]
    for i in range(len(z1)): result[i + m] += z1[i]
    for i in range(len(z2)): result[i + 2 * m] += z2[i]

    return result[:2 * n - 1]

File: 18_multiplication_linalg/poly.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void add_poly(int* a, int* b, int* res, int n) {
    for (int i = 0; i < n; i++)
        res[i] = a[i] + b[i];
}

void sub_poly(int* a, int* b, int* res, int n) {
    for (int i = 0; i < n; i++)
        res[i] = a[i] - b[i];
}

void poly_multiply(int* p, int* q, int* res, int n) {
    if (n == 1) {
        res[0] = p[0] * q[0];
        return;
    }

    int m = n / 2;
    int *low_p = p;
    int *high_p = p + m;
    int *low_q = q;
    int *high_q = q + m;

    int* z0 = calloc(2 * m, sizeof(int));
    int* z1 = calloc(2 * m, sizeof(int));
    int* z2 = calloc(2 * m, sizeof(int));
    int* sum_p = calloc(m, sizeof(int));
    int* sum_q = calloc(m, sizeof(int));
    int* temp1 = calloc(2 * m, sizeof(int));

    poly_multiply(low_p, low_q, z0, m);
    poly_multiply(high_p, high_q, z2, m);

    add_poly(low_p, high_p, sum_p, m);
    add_poly(low_q, high_q, sum_q, m);
    poly_multiply(sum_p, sum_q, temp1, m);
    sub_poly(temp1, z0, temp1, 2 * m);
    sub_poly(temp1, z2, z1, 2 * m);

    memset(res, 0, sizeof(int) * 2 * n);
    for (int i = 0; i < 2 * m; i++) res[i] += z0[i];
    for (int i = 0; i < 2 * m; i++) res[i + m] += z1[i];
    for (int i = 0; i < 2 * m; i++) res[i + 2 * m] += z2[i];

    free(z0); free(z1); free(z2); free(sum_p); free(sum_q); free(temp1);
}